PWGCF/TwoParticleCorrelations/Tasks/r2p2-4-id.cxx

```diff
@@ -0,0 +1,444 @@
+  {
+    int mult = 0;
+    int pid1, pid2;
+    int sign1, sign2;
+    int etabin1, phibin1, etabin2, phibin2;
+
+    for (auto track1 : tracks) {
+      mult++;
+      histos.fill(HIST("h1d_n1_phi"), track1.phi());
+      histos.fill(HIST("h1d_n1_eta"), track1.eta());
+      histos.fill(HIST("h1d_n1_pt"), track1.pt());
+      //---Single Particle Distribution----------------------------------------
+      sign1 = (track1.sign() + 1) / 2;
+      pid1 = track1.pid();
+      for (int i = pid1;; i /= 10) // Filling histograms for different particle species; i%10 = 0 means ch, 1->pi, 2->ka, 3->pr.
+      {
+        hist.h1d_1p[i % 10][sign1]->Fill(track1.pt(), 1.0 / (2.0 * constants::math::PI * track1.pt())); // h1d_n1_pt*
+        hist.h2d_1p[i % 10][sign1][0]->Fill(track1.eta(), track1.phi());                                // h2d_n1_etaphi*
+        hist.h2d_1p[i % 10][sign1][1]->Fill(track1.eta(), track1.phi(), track1.pt());                   // h2d_pt_etaPhi*
+        if (i < 1)
+          break;
+      }
+      //-----------------------------------------------------------------------
+      etabin1 = (track1.eta() + 0.8) * 15; // 15= 24/1.6
+      phibin1 = 36 * track1.phi() / (2 * constants::math::PI);
+
+      for (auto track2 : tracks) {
+        if (track1.index() == track2.index())
+          continue;
+        etabin2 = (track2.eta() + 0.8) * 15; // 15= 24/1.6
+        phibin2 = 36 * track2.phi() / (2 * constants::math::PI);
+        sign2 = (track2.sign() + 1) / 2;
+        pid2 = track2.pid();
+        //-----Two Particle Distribution---------------------------------------
+        int i, n1, n2;
+        for (int e = pid1;; e /= 10) {
+          n1 = e % 10;
+
+          for (int f = pid2; (f % 10) >= n1; f /= 10) {
+            if ((e == 0) && (f != 0))
+              continue; // No ch-pi or ch-ka or ch-pr pairs
+
+            n2 = f % 10;
+            i = ((n2 - n1) / 2 + (n2 - n1) % 2);
+            i = i + (n1 + n2) * (1 + i) / 2; // This formula gives 0 for chch, 1 for pipi, 2 for kaka, 3->prpr, 4->pik, 5->pip, 6->kp
+
+            hist.h2d_2p[i][sign1][sign2][0]->Fill(36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5);                            // h2d_n2_eta1Phi1Eta2Phi2*
+            hist.h2d_2p[i][sign1][sign2][1]->Fill(36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track2.pt());               // h2d_npt_eta1Phi1Eta2Phi2*
+            hist.h2d_2p[i][sign1][sign2][2]->Fill(36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt());               // h2d_ptn_eta1Phi1Eta2Phi2*
+            hist.h2d_2p[i][sign1][sign2][3]->Fill(36 * etabin1 + phibin1 + 0.5, 36 * etabin2 + phibin2 + 0.5, track1.pt() * track2.pt()); // h2d_ptpt_eta1Phi1Eta2Phi2*
+            if (f < 1)
+              break;
+          }
+          if (e < 1)
+            break;
```

Comment:
This continues being too much processing for being in the double track inner loop

Consider the following tagging/indexing
```
  kCharged, /* reco track NOT identified as pi, ka, nor pr */
  kPion,    /* reco identified as pi */
  kKaon,    /* reco identified as ka */
  kProton,  /* reco identified as pr */
```
when BOTH tracks ARE `kCharged` only the ch-ch associated histograms are filled
when BOTH trakcs ARE NOT `kCharged`, after filling the corresponding id1-id2 histograms, the ch-ch associated histograms are additionally filled

Does it make sense?

Reply:
Yes I understand what you're saying. The reason why I had to use the double loop for filling the histograms is that, as I mentioned above, a track can be identified as more than one kind of particle with the current PID technique. Thus in one iteration of the inner track loop, I may have to fill several histograms of different combinations of particles. Suppose, as the worst case,  particle 1 and particle 2 are both identified to be pion, kaon, and proton, then in the inner track loop the histograms need to be filled for the 7 different combinations of particles. 
I had planned that after improving the PID, where a track is identified as only one kind of particle, I would drop the double loop and use an algorithm similar to what you suggested now.